File: examples/kotlin/src/main/kotlin/com/xebia/functional/xef/auto/ChessAI.kt
package com.xebia.functional.xef.auto

import com.xebia.functional.xef.auto.llm.openai.conversation
import com.xebia.functional.xef.auto.llm.openai.prompt
import kotlinx.serialization.Serializable

@Serializable data class ChessMove(val player: String, val move: String)

@Serializable data class ChessBoard(val board: String)

@Serializable data class GameState(val ended: Boolean, val winner: String)

suspend fun main() {
  conversation {
    val moves = mutableListOf<ChessMove>()
    var gameEnded = false
    var winner = ""

    while (!gameEnded) {
      val currentPlayer = if (moves.size % 2 == 0) "Player 1 (White)" else "Player 2 (Black)"
      val prompt =
        """
              |$currentPlayer, it's your turn. 
              |Previous moves: ${moves.joinToString(", ")}
              |Make your next move:
          """
          .trimIndent()

      val move: ChessMove = prompt(prompt)
      moves.add(move)

      // Update boardState according to move.move
      // ...

      val boardPrompt =
        """
              Given the following chess moves: ${moves.joinToString(", ") { it.player + ":" + it.move }}},
              generate a chess board on a table with appropriate emoji representations for each move and piece.
              Add a brief description of the move and it's implications
          """
          .trimIndent()

      val chessBoard: ChessBoard = prompt(boardPrompt)
      println("Current board:\n${chessBoard.board}")

      val gameStatePrompt =
        """
              Given the following chess moves: ${moves.joinToString(", ")},
              has the game ended (win, draw, or stalemate)?
          """
          .trimIndent()

      val gameState: GameState = prompt(gameStatePrompt)

      gameEnded = gameState.ended
      winner = gameState.winner
    }

    println("Game over. Final move: ${moves.last()}, Winner: $winner")
  }
}

File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/auto/ChessAI.scala
package com.xebia.functional.xef.scala.auto

import com.xebia.functional.xef.scala.auto.*
import io.circe.Decoder

import scala.annotation.tailrec

private final case class ChessMove(player: String, move: String) derives SerialDescriptor, Decoder

private final case class ChessBoard(board: String) derives SerialDescriptor, Decoder

private final case class GameState(ended: Boolean, winner: Option[String]) derives SerialDescriptor, Decoder

@tailrec
private def chessGame(moves: List[ChessMove], gameState: GameState)(using conversation: ScalaConversation): (String, ChessMove) =
  if !gameState.ended then
    val currentPlayer = if moves.size % 2 == 0 then "Player 1 (White)" else "Player 2 (Black)"

    val previousMoves = moves.map(m => m.player + ":" + m.move).mkString(", ")

    val movePrompt = moves match {
      case Nil => s"""
                     |$currentPlayer, you are playing chess and it's your turn.
                     |These are no previous chess board moves.
                     |Make your first move:
      """.stripMargin
      case l => s"""
                   |$currentPlayer, you are playing chess and it's your turn.
                   |These are the previous chess board moves: $previousMoves
                   |Make your next move:
      """.stripMargin
    }
    println(movePrompt)
    val move: ChessMove = prompt(movePrompt)
    println(s"Move is: $move")

    val boardPrompt =
      s"""
         |Given the following chess moves: $previousMoves,
         |generate a chess board on a table with appropriate emoji representations for each move and piece.
         |Add a brief description of the move and it's implications
      """.stripMargin

    val chessBoard: ChessBoard = prompt(boardPrompt)
    println(s"Current board:\n${chessBoard.board}")

    val gameStatePrompt =
      s"""
         |Given the following chess moves: ${moves.mkString(", ")},
         |has the game ended (win, draw, or stalemate)?
      """.stripMargin

    val gameState: GameState = prompt(gameStatePrompt)

    chessGame(moves :+ move, gameState)
  else (gameState.winner.getOrElse("Something went wrong"), moves.last)

@main def runChess: Unit =
  conversation {
    val (winner, fMove) = chessGame(Nil, GameState(false, None))
    println(s"Game over. Final move: $fMove, Winner: $winner")
  }
